src/com/googlecode/jmxtrans/model/output/GraphiteWriter.java

```diff
@@ -101,6 +101,11 @@ public void internalWrite(Server server, Query query, ImmutableList<Result> resu
 							log.debug("Graphite Message: {}", line);
 							writer.write(line);
 							writer.flush();
+							if (writer.checkError()) {
+								log.error("Error writing to Graphite, clearing Graphite socket pool");
+								pool.returnObject(address, socket);
```

Comment:
Actually, your implementation is robust. You do get a new connection after returning the faulty one. Still, we should use `invalidateObject()` instead of `clear()`.